src/model.ts

```diff
@@ -90,23 +123,52 @@ export abstract class Model<T extends Model<T>> {
       await data.remove(this.__snapshot)
     }
 
-    const labels: {[label in Labels]?: string} = {}
-    for (const key of keys) {
-      if (isSecondary(key.index)) {
-        labels[key.index.options.label] = key.query()
+    await data.set(primary, snakecase(prune(this), { deep: true }), this.secondaries())
+
+    if (this.shadowKeys) {
+      if (
+        this.shadowKeys().length > MAX_SHADOW_KEYS &&
+        !this.unsafeShadowKeysUnbounded?.()
```

Comment:
I think this would not work as intended:

- This will ignore the result from `this.unsafeShadowKeysUnbounded()`, while when trying to get an unbounded number of shadow keys, I guess the intuitive thing would be to just override this function.

- This will assume that there are no shadow keys if `shadowKeys()` method is not defined. However, one might only defined `unsafeShadowKeysUnbounded()` instead. In fact, I feel the intuitive default is to override only one of the two and not both (you either define shadow keys normally, or you choose the unsafe, unbounded way).

My recommendation would be to have another intermediary method who figures out the shadow keys (and checks the bounds as well), and calling that function in `save()`, separating the _bounding_ logic into another method.

```ts
resolveShadowKeys() {
  if (this.unsafeShadowKeysUnbounded) {
    return this.unsafeShadowKeysUnbounded()
  } else if (this.shadowKeys) {
    const res = this.shadowKeys()
    if (res.length > MAX_SHADOW_KEYS) {
      throw ...
    }
    
    return res
  }
}
```

Reply:
So I think the way I have implemented the unsafeShadowKeysUnbounded is different to how you expected, essentially it is just a flag to say to ignore the MAX_SHADOW_KEYS restriction, the shadow keys are always set in the shadowKeys method. I was thinking that it might make more sense to add an options method instead:
```
setOptions() {
    return {
      shadowKeysLimit: false
    }
 }
```

Reply:
In that case I would highly recommend switching to having one overriden method whose name acts as the flag (either `shadowKeys()` or `unsafeShadowKeysUnbounded()`), instead of one or two overriden methods where one is a flag determining how the other is treated (both `shadowKeys()` and `shadowKeysBoundFlag()` or `setOptions()` or another method). 


In the first pattern (which I believe is [somewhat familiar already](https://reactjs.org/docs/react-component.html#unsafe_componentwillupdate)), the behavior is isolated and clear (here we have shadow keys, or here we have unbounded shadow keys which is potentially unsafe). In the second pattern, you would need to look at another part of the code (or potentially some other class entirely) to understand the behavior, which increases complexity.

Reply:
Would it be better to change `unsafeShadowKeysUnbounded` to `UNSAFE_shadowKeysUnbounded`? 

If keys are provided for both `unsafeShadowKeysUnbounded` and `shadowKeys` methods, would it make sense to throw an error, or just document that in this case any keys provided to the `shadowKeys` method will be ignored?

Reply:
Good idea, `UNSAFE_shadowKeysUnbounded()` better communicates the underlying unsafety, so it indeed is a better choice.

I would recommend throwing an error for now, since that restriction can be relaxed later without breaking backwards compatibility, but not the other way around.